### PR TITLE
bpo-32321: Add pure Python fallback for functools.reduce

### PR DESCRIPTION
**Pull request by bradengroom**
https://bugs.python.org/issue32321

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately our records indicate you have not signed the CLA. For legal reasons we need you to sign this before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for your contribution, we look forward to reviewing it!

File: Lib/functools.py [hunk 1]
```diff
@@ -16,7 +16,15 @@
 try:
     from _functools import reduce
 except ImportError:
-    pass
+    def reduce(function, iterable, initializer=None):
+        it = iter(iterable)
+        if initializer is None:
+            value = next(it)
```
Comment (eric-wieser) [inline Lib/functools.py]: This raises `StopIteration` when it should raise `TypeError: reduce() of empty sequence with no initial value`

**Review by eric-wieser — CHANGES_REQUESTED** [review]

**Comment by bradengroom** [comment]
@eric-wieser Realized I was missing some other pieces here. My fixup commit tests the new implementation as well as the C module. It also should raise the same exceptions as the C implementation now.

File: Lib/functools.py [hunk 2]
```diff
@@ -16,7 +16,34 @@
...
+            from left to right, so as to reduce the sequence to a single value.
+            For example, reduce(lambda x, y: x+y, [1, 2, 3, 4, 5]) calculates
+            ((((1+2)+3)+4)+5).  If initial is present, it is placed before the items
+            of the sequence in the calculation, and serves as a default when the
+            sequence is empty.
+        """
+        try:
+            iterable = iter(iterable)
+        except TypeError as err:
+            raise TypeError('reduce() arg 2 must support iteration')
```
Comment (pppery) [inline Lib/functools.py]:
```suggestion
            raise TypeError('reduce() arg 2 must support iteration') from None
```
You don't want the first TypeError to show up in the traceback

File: Lib/functools.py [hunk 3]
```diff
@@ -16,7 +16,34 @@
...
+
+        if initializer is _initial_missing:
+            try:
+                value = next(iterable)
+            except StopIteration:
+                raise TypeError("reduce() of empty sequence with no initial value")
```
Comment (pppery) [inline Lib/functools.py]:
```suggestion
                raise TypeError("reduce() of empty sequence with no initial value") from None
```
See above

**Comment by bradengroom** [comment]
@pppery Thanks
https://github.com/python/cpython/pull/9884/commits/63bfc75a1973801b6faca4b3e67c4cbb009fa115

**Comment by bradengroom** [comment]
@rhettinger This is ready for core review. It looks like you are a code owner for this file. What's the proper way to ping for review?

**Comment by vstinner** [comment]
Oh. I wasn't aware that there were two PRs implementing the same thing. Sorry, I just merged the PR #8548.

**Comment by bradengroom** [comment]
Oh I missed the other PR in the issue tracker somehow. Thanks

**Pull request closed without merging**